flink-state-backends/flink-statebackend-heap-spillable/src/test/java/org/apache/flink/runtime/state/heap/CopyOnWriteSkipListStateMapTest.java

```diff
@@ -115,577 +144,783 @@ public void testInitStateMap() {
+				assertEquals("11", stateMap.putAndGetOld(key, namespace, newValue));
+				addToReferenceState(referenceStates, key, namespace, newValue);
+				assertEquals("111", stateMap.get(key, namespace));
+				assertTrue(stateMap.containsKey(key, namespace));
+				if (withSnapshot) {
+					snapshot.release();
+				}
+				return new Tuple3<>(totalSize, totalSize, totalSpaceNumber);
+			}
+		);
+	}
 
-		// remove (key 4, namespace 4), and it should be logically removed
-		assertEquals("4", stateMap.removeAndGetOld(4, 4L));
-		removeFromReferenceState(referenceStates, 4, 4L);
-		// a space should be allocated
-		totalStateSize--;
-		totalLogicallyRemovedKey++;
-		totalSpaceNumber += 1;
-		assertEquals(totalSpaceNumber, spaceAllocator.getTotalSpaceNumber());
-		assertNull(stateMap.get(4, 4L));
-		assertFalse(stateMap.containsKey(4, 4L));
-		assertEquals(totalStateSize, stateMap.size());
-		assertEquals(totalSizeIncludingLogicalRemovedKey, stateMap.totalSize());
-		assertEquals(totalLogicallyRemovedKey, stateMap.getLogicallyRemovedNodes().size());
-		verifyState(referenceStates, stateMap);
+	/**
+	 * Test put -> remove without snapshot.
+	 */
+	@Test
+	public void testPutAndRemove() {
+		testPutAndRemove(false);
+	}
 
-		// remove (key 4, namespace 4) again, and nothing should happen
-		assertNull(stateMap.removeAndGetOld(4, 4L));
-		assertEquals(totalSpaceNumber, spaceAllocator.getTotalSpaceNumber());
-		assertEquals(totalStateSize, stateMap.size());
-		assertEquals(totalSizeIncludingLogicalRemovedKey, stateMap.totalSize());
-		assertEquals(totalLogicallyRemovedKey, stateMap.getLogicallyRemovedNodes().size());
-		verifyState(referenceStates, stateMap);
+	/**
+	 * Test put -> remove during snapshot, put should trigger copy-on-write and remove shouldn't.
+	 */
+	@Test
+	public void testPutAndRemoveWithSnapshot() {
+		testPutAndRemove(true);
+	}
 
-		// put the logically removed (key 4, namespace 4)
-		assertNull(stateMap.putAndGetOld(4, 4L, "44"));
-		addToReferenceState(referenceStates, 4, 4L, "44");
-		totalStateSize++;
-		totalLogicallyRemovedKey--;
-		assertEquals(totalSpaceNumber, spaceAllocator.getTotalSpaceNumber());
-		assertEquals("44", stateMap.get(4, 4L));
-		assertTrue(stateMap.containsKey(4, 4L));
-		assertEquals(totalStateSize, stateMap.size());
-		assertEquals(totalSizeIncludingLogicalRemovedKey, stateMap.totalSize());
-		assertEquals(totalLogicallyRemovedKey, stateMap.getLogicallyRemovedNodes().size());
-		verifyState(referenceStates, stateMap);
+	private void testPutAndRemove(boolean withSnapshot) {
+		final int key = 6;
+		final long namespace = 6L;
+		final String value = "66";
+		testWithFunction(
+			(totalSize, stateMap, referenceStates) -> {
+				int totalLogicallyRemovedKey = 0;
+				int totalSizeIncludingLogicalRemovedKey = totalSize;
+				CopyOnWriteSkipListStateMapSnapshot<Integer, Long, String> snapshot = null;
+				if (withSnapshot) {
+					// take a snapshot
+					snapshot = stateMap.stateSnapshot();
+				}
```

Comment:
Instead of pushing all logic deep into the one uber test function I think it is better to use a compositional approach. E.g. one could write the test with `withSnapshot = false` like this

```
putAndRemove()
makeAssertions()
```

and the one with `withSnapshot == true`

```
takeSnapshot()
putAndRemove()
releaseSnapshot()
makeAssertions()
```